src/meta/src/rpc/server.rs

```diff
@@ -171,308 +158,141 @@ pub async fn rpc_serve_with_store<S: MetaStore>(
                 leader_addr.host,
                 leader_addr.port
             );
-
-            if note_status_leader_rx.changed().await.is_err() {
-                panic!("Issue receiving leader value from channel");
-            }
         }
     });
 
-    // FIXME: Start leader services if follower becomes leader
-    // failover logic
+    // FIXME: maybe do not use a channel here
+    // What I need is basically a oneshot channel with one producer and multiple consumers
+    let (svc_shutdown_tx, mut svc_shutdown_rx) = WatchChannel(());
 
-    // Start follower services
-    if !leader_rx.borrow().1 {
-        tracing::info!("Node initially elected as follower");
-        tokio::spawn(async move {
-            let span = tracing::span!(tracing::Level::INFO, "services");
-            let _enter = span.enter();
+    let join_handle = tokio::spawn(async move {
+        let span = tracing::span!(tracing::Level::INFO, "services");
+        let _enter = span.enter();
 
-            let health_srv = HealthServiceImpl::new();
-            // run follower services until node becomes leader
-            tracing::info!("Starting follower services");
+        // failover logic
+        services_leader_rx
+            .changed()
+            .await
+            .expect("Leader sender dropped");
+
+        let is_leader = services_leader_rx.borrow().clone().1;
+        let was_follower = !is_leader;
+
+        // run follower services until node becomes leader
+        // FIXME: Add service discovery for follower
+        // https://github.com/risingwavelabs/risingwave/issues/6755
+        let mut svc_shutdown_rx_clone = svc_shutdown_rx.clone();
+        let (follower_shutdown_tx, follower_shutdown_rx) = OneChannel::<()>();
+        let (follower_finished_tx, follower_finished_rx) = OneChannel::<()>();
```

Comment:
We really need `follower_finished_tx` here? Can we just expose this handler and await at L239?

Reply:
Thank you for pointing this out. Will definitely use a join handle here.